File: engine/argot/scoring/scorers/typical_call_density.py
"""Typical-call-density shape primitive.

Measures under-coverage of cluster-typical callees: the fraction of
``call_expression`` nodes in a hunk whose callee is among the cluster's
top-10 most-attested callees.

Per-cluster baseline: ``(top10_set, mean, std)`` of per-file typical-call
density across the cluster's files.  Only callees that are NOT ``None``
(i.e. whose AST callee chain resolves to a dotted name) contribute to the
numerator; all call-expression nodes — including those whose callee cannot
be resolved — contribute to the denominator.  This is intentional: a hunk
whose calls bottom out at subscripts or string literals has density 0, which
is the anomaly that fires the primitive.

Score-time: one-sided tail-z (positive when hunk is below cluster mean,
i.e. the hunk under-uses cluster-typical callees).
Ramp: ``min(cluster_bonus_clip, max(0.0, z - 1.0))``.

Callee-frequency convention: top-10 is ranked by per-file presence count
(how many cluster files contain each callee), not by raw occurrence count.
This is consistent with ``cluster_callee_counts`` in
``CallReceiverScorer._build_clusters``.

Asymmetric-by-construction: cal hunks come from ``model_a_files`` attached
to typical functions that call cluster-typical things, so their density sits
at or above the cluster mean.  Only hunks that structurally avoid
cluster-typical callees (e.g. string-formula synthesis) fire the primitive.
"""

from __future__ import annotations

import math
from collections import Counter
from collections.abc import Iterable
from dataclasses import dataclass
from pathlib import Path

from argot.scoring.scorers.call_receiver import extract_callees
from argot.scoring.scorers.shape_primitive import Language

# Number of most-frequent cluster callees used as the "typical-call set".
_TOP_N: int = 10

# Minimum number of cluster files that must contribute a defined density
# for the per-cluster baseline to be trusted.
_MIN_VALID_FILES: int = 3


@dataclass(frozen=True)
class _TypicalCallDensityBaseline:
    top10_set: frozenset[str]
    mean: float
    std: float


def _compute_density(
    source: str,
    language: Language,
    top10_set: frozenset[str],
) -> float | None:
    """Return fraction of call nodes whose callee is in top10_set.

    Returns ``None`` when the source has 0 call-expression nodes (no
    denominator — the ratio is undefined; score-time abstains on None).
    ``None`` callees count toward the denominator but never toward the
    numerator.
    """
    callees = extract_callees(source, language)
    denom = len(callees)
    if denom == 0:
        return None
    hits = sum(1 for c in callees if c is not None and c in top10_set)
    return hits / denom


class TypicalCallDensity:
    """Scalar shape primitive: typical-call-density under-coverage.

    Implements ``ShapePrimitive[_TypicalCallDensityBaseline]``.
    Language is not known at construction time; it is captured on the first
    ``fit_cluster_baseline`` call and reused at score time.
    """

    name: str = "typical_call_density"
    min_cluster_size: int = 10
    cluster_bonus_clip: float = 5.0

    def __init__(self) -> None:
        self._language: Language | None = None

    def fit_cluster_baseline(
        self,
        cluster_files: Iterable[tuple[Path, str]],
        language: Language,
    ) -> _TypicalCallDensityBaseline | None:
        """Compute the top-10 callee set and (mean, std) of per-file densities.

        Callee frequency is measured by per-file presence count: each callee
        that appears in a cluster file contributes 1 to its count regardless
        of how many times it is called within that file.  ``None`` callees
        (unresolvable callee chains) are excluded from frequency counting.

        Skips files with 0 call-expression nodes (undefined denominator).
        Returns ``None`` when fewer than ``_MIN_VALID_FILES`` files contribute
        a density.
        """
        self._language = language

        files: list[tuple[Path, str]] = list(cluster_files)

        # Build per-file presence counts for each callee across the cluster.
        callee_file_counts: Counter[str] = Counter()
        for _path, source in files:
            seen: set[str] = set()
            for callee in extract_callees(source, language):
                if callee is not None and callee not in seen:
                    callee_file_counts[callee] += 1
                    seen.add(callee)

        # Top-N by per-file frequency; frozenset for O(1) membership test.
        top10_set: frozenset[str] = frozenset(c for c, _ in callee_file_counts.most_common(_TOP_N))

        # Per-file densities; skip files with 0 call nodes.
        densities: list[float] = []
        for _path, source in files:
            d = _compute_density(source, language, top10_set)
            if d is not None:
                densities.append(d)

        if len(densities) < _MIN_VALID_FILES:
            return None

        mean = sum(densities) / len(densities)
        variance = sum((d - mean) ** 2 for d in densities) / len(densities)
        std = math.sqrt(variance)
        return _TypicalCallDensityBaseline(top10_set=top10_set, mean=mean, std=std)

    def score(
        self,
        hunk_content: str,
        *,
        baseline: _TypicalCallDensityBaseline | None,
        cluster_size: int,
    ) -> float:
        """Return one-sided tail-z contribution, clipped at ``cluster_bonus_clip``.

        Positive z signals under-usage of cluster-typical callees.
        Ramp: ``min(clip, max(0.0, z - 1.0))``.

        Abstains (returns 0.0) when:
        - ``baseline`` is None (cluster had too few files with calls),
        - ``cluster_size < self.min_cluster_size`` (cluster too small),
        - language was never set (fit was never called), or
        - the hunk has 0 call-expression nodes, or
        - the top-10 callee set is empty (cluster had no resolvable callees).
        """
        if baseline is None or cluster_size < self.min_cluster_size:
            return 0.0
        language = self._language
        if language is None:
            return 0.0
        if not baseline.top10_set:
            return 0.0
        hunk_density = _compute_density(hunk_content, language, baseline.top10_set)
        if hunk_density is None:
            return 0.0
        z = (baseline.mean - hunk_density) / max(baseline.std, 1e-6)
        return min(self.cluster_bonus_clip, max(0.0, z - 1.0))


__all__ = ["TypicalCallDensity"]

File: crates/argot-rules-voice/src/scoring/shape_primitives/typical_call_density.rs
//! Typical-call-density under-coverage.
//!
//! Per-file density = (#call nodes whose non-`None` callee is in the cluster's
//! top-10) / (total call nodes, including unresolved callees). The top-10 is
//! ranked by per-file PRESENCE count with first-insertion tie-break (document
//! order across files), matching Python's `Counter.most_common(10)`.
//! One-sided tail-z ramp (fires only on under-coverage).

use crate::scoring::adapters::Language;
use crate::scoring::call_receiver::extract_callees;
use crate::scoring::shape_primitive::{Baseline, ShapePrimitive};
use crate::scoring::shape_primitives::population_mean_std;
use std::collections::{BTreeSet, HashMap, HashSet};
use std::path::PathBuf;
use std::sync::OnceLock;

const TOP_N: usize = 10;
const MIN_VALID_FILES: usize = 3;

/// Fraction of call nodes whose callee is in `top10_set`. `None` when the
/// source has 0 call nodes (undefined denominator). Unresolved (`None`)
/// callees count toward the denominator but never the numerator.
fn compute_density(source: &str, language: Language, top10_set: &BTreeSet<String>) -> Option<f64> {
    let callees = extract_callees(source, language);
    let denom = callees.len();
    if denom == 0 {
        return None;
    }
    let hits = callees
        .iter()
        .filter(|c| c.as_deref().map(|s| top10_set.contains(s)).unwrap_or(false))
        .count();
    Some(hits as f64 / denom as f64)
}

/// Typical-call-density under-coverage primitive. Language captured on fit.
#[derive(Default)]
pub struct TypicalCallDensity {
    language: OnceLock<Language>,
}

impl ShapePrimitive for TypicalCallDensity {
    fn name(&self) -> &str {
        "typical_call_density"
    }
    fn min_cluster_size(&self) -> usize {
        10
    }
    fn cluster_bonus_clip(&self) -> f64 {
        5.0
    }

    fn set_language(&self, language: Language) {
        let _ = self.language.set(language);
    }

    fn fit_cluster_baseline(
        &self,
        cluster_files: &[(PathBuf, String)],
        language: Language,
    ) -> Option<Baseline> {
        let _ = self.language.set(language);

        // Per-file presence counts, preserving first-insertion order so the
        // top-10 tie-break matches Counter.most_common.
        let mut order: Vec<String> = Vec::new();
        let mut counts: HashMap<String, u64> = HashMap::new();
        for (_path, source) in cluster_files {
            let mut seen: HashSet<String> = HashSet::new();
            for callee in extract_callees(source, language).into_iter().flatten() {
                if seen.insert(callee.clone()) {
                    match counts.get_mut(&callee) {
                        Some(v) => *v += 1,
                        None => {
                            counts.insert(callee.clone(), 1);
                            order.push(callee);
                        }
                    }
                }
            }
        }

        // most_common(TOP_N): stable sort by count desc, first-insertion for
        // ties (explicit index tiebreak), take TOP_N.
        let mut entries: Vec<(usize, &String, u64)> = order
            .iter()
            .enumerate()
            .map(|(i, c)| (i, c, counts[c]))
            .collect();
        entries.sort_by(|a, b| b.2.cmp(&a.2).then(a.0.cmp(&b.0)));
        let top10_set: BTreeSet<String> = entries
            .iter()
            .take(TOP_N)
            .map(|(_, c, _)| (*c).clone())
            .collect();

        let mut densities: Vec<f64> = Vec::new();
        for (_path, source) in cluster_files {
            if let Some(d) = compute_density(source, language, &top10_set) {
                densities.push(d);
            }
        }
        if densities.len() < MIN_VALID_FILES {
            return None;
        }
        let (mean, std) = population_mean_std(&densities);
        Some(Baseline::Top10MeanStd {
            top10_set,
            mean,
            std,
        })
    }

    fn score(&self, hunk: &str, baseline: Option<&Baseline>, cluster_size: usize) -> f64 {
        let Some(Baseline::Top10MeanStd {
            top10_set,
            mean,
            std,
        }) = baseline
        else {
            return 0.0;
        };
        if cluster_size < self.min_cluster_size() {
            return 0.0;
        }
        let Some(language) = self.language.get().copied() else {
            return 0.0;
        };
        if top10_set.is_empty() {
            return 0.0;
        }
        let Some(hunk_density) = compute_density(hunk, language, top10_set) else {
            return 0.0;
        };
        // One-sided: positive z = hunk below cluster mean = under-coverage.
        let z = (*mean - hunk_density) / std.max(1e-6);
        (z - 1.0).max(0.0).min(self.cluster_bonus_clip())
    }
}
